### DISCUSSION TITLE
[DOCS]: Add documentation for configuring profiles

### DISCUSSION THREAD

**Discussion by rjmackay** [sensedeep/onetable-cli#7] [Q&A]
**Describe the bug**
There is no documentation for how to configure profiles.

**Additional context**
Docs tell you how to pick a profile, but provide no context to what that does. From reading the code it looks like maybe you can set `profiles` in `migrate.json` but I'm not sure how best to use that?

**Comment by mobsense** [comment]
Sure, thank you. We'll add.

Profiles are pretty simple. When loading, the map for a profile is hoisted to the top level. Here is one migrate.json we use:

If the profile is set to 'dev', the [dir, name, endpoint] properties are copied to the root level.

```
{
    config: ['../pak.json', '../product.json', '../CONFIG/service-keys.json'],
    profiles: {
        dev: {
            dir: './migrations',
            name: 'sensedb',
            endpoint: 'http://localhost:8000'
        },
        qa: {
            name: 'sensedb',
            arn: 'arn:aws:lambda:us-east-1:xxxx:function:migrate-qa-invoke'
        },
        prod: {
            name: 'sensedb',
            arn: 'arn:aws:lambda:us-east-1:xxxx:function:migrate-prod-invoke'
        }
    }
}

```

**Reply by rjmackay** [reply]
So the config format itself is pretty confusing. It took a lot of trial and error to figure out what needed to be under `onetable` or not, and I had to define the migration dir twice in `dir` (for Migrate) and `migrations` (for generateMigration).
If you define `migrations` at the top level it gets passed into `Migrate` which breaks because it's expecting a list of migrations. 

I've ended up with this
```
{
  "onetable": {
    "schema": "src/data/db/schema.ts",
    "name": "aje-offering",
    "isoDates": true,
    "aws": {
      "region": "us-east-1"
    },
    "migrations": "migrations/"
  },
  "dir": "migrations/",
  "profiles": {
    "local": {
      "onetable": {
        "aws": {
          "endpoint": "http://localhost:4567"
        }
      }
    },
    "dev": {},
  }
}
```

**Comment by rjmackay** [comment]
Awesome. Thank you!

**Comment by mobsense** [comment]
Let me know how you go with migrations in general. We find them vital to our workflow. Just setup discussion on this repo for this purpose. So I'll move this to discussions

**Comment by rjmackay** [comment]
Would it make sense to add an option to auto-create the table before migrations runs? ie. when running locally

**Reply by mobsense** [reply]
Help me understand the scenario.  Is this for unit tests or general dev.

**Reply by rjmackay** [reply]
For local dev and testing migrations. For unittest we use onetables `table.createTable()` to automate creating the table in dynamodb-local. I'd like to do similar for local dev.

The quick workaround is just to add a quick script that does this for me. But it seems like something that the CLI could easily do since it already has the schema.

**Reply by mobsense** [reply]
Cool. Thank you.

**Comment by rjmackay** [comment]
Also curious if you've run migrations in TS at all? Our codebases are entirely TS so it'd be good to do that same with migrations

**Reply by mobsense** [reply]
Our migrations are JS, but I'm really keen and will be responsive to support any TS issue.

The migrateController is a very small piece of code really (~1 page).

**Reply by rjmackay** [reply]
FWIW `node --loader ts-node/esm node_modules/.bin/onetable` seems to works perfectly so far. I'm defining the schema in TS and migrations too

**Reply by rjmackay** [reply]
Ah correction: Schema in TS works, but migrations don't because the import always looks for a `.js` file

**Reply by mobsense** [reply]
For the local case, you will need to have transpiled your TS code. So won't you have a JS to point to?

For the remote case, you will have referenced the imports of them via the Migrate.migrations constructor property.

**Reply by rjmackay** [reply]
> For the local case, you will need to have transpiled your TS code. So won't you have a JS to point to?

When running through ts-node ie. `node --loader ts-node/esm` it transpiles on the fly, so there's not JS file output to disk.

**Reply by mobsense** [reply]
I'm not familiar with that little bit of loader magic. How do you integrate that with the CLI?

If you modify onetable-migrate index.js in the loadMigration to load .ts files, does it work?  Is it that simple?

Could you explain a bit more how you integrate with the CLI?  Do you plan to use the remote migrate or do you only plan to use local migrations?

**Reply by rjmackay** [reply]
Only just coming back to this sorry. So you have to have `ts-node` installed, then I was running `node --loader ts-node/esm node_modules/.bin/onetable` to call the CLI.
I'd initially just been planning to use local migrations, possibly running from CI. But I haven't gotten very deep into it yet to know what makes the most sense.

**Comment by rjmackay** [comment]
@mobsense I might be missing something, but it looks like the cli looks for the migrations directory in the config `migrations` sometimes and `dir` others?

**Reply by mobsense** [reply]
Should be dir. I noticed 2 places in the cli that refer to config.migrations or onetable.migrations. These should be config.dir.

The Migrate({migrations}) property is to provide an in-memory array of migrations. When working locally, you can have a directory of migration scripts that are loaded. When working to a remote DynamoDB table, you need a proxy lambda that the CLI talks to. The onetable-migrate README has an example of that simple lambda.

When working remotely, the CLI uses the arn property in migrate.json to specify the lambda. The migrations are all loaded into the lambda and configured via Migrate.migrations.

**Reply by rjmackay** [reply]
Right. As mentioned above the main place I found config.migrations being used was generateMigration.

**Comment by rjmackay** [comment]
One thing that would also be really helpful is some examples of complex migrations.

**Reply by mobsense** [reply]
We've got a large library of these, having used migrations for about 12 months now.

What we find is that we've developed a library of helper routines that our migrations use. These helpers do things like look for unused attributes, items, remove closed accounts, update items to update GSI indexes, etc.

Generally, we try to keep migrations really small and discrete and break up complex migrations into multi-steps. We often do backups between each step to lower risk.  We're going to add backup snapshots to the cli for this purpose.

The migrations themselves fall into two categories.

1. Really simple mutations like adding an attribute to existing items. Removing old items etc.
2.More complex mutations that need to refactor existing items and either create other new entities or change data values / types in a multi-part way.

For (2) we often find migrations need to be multi-step.

1. Update code to handle old and new. This includes updating the schema without breaking backward compatibility for old data and old code.
2. Run migration to populate the new schema elements
3. Update code to remove old logic.
4. Run another migration to cleanup.

generally, to use a climbing analogy ... make then break.

What also really helps is adding defensive code to the backend to handle cases during data migration so you can handle old and new format data while the migration takes place.

Here is a simplified migration to update a Member entity GSIs

In this example, we use scan whereas we'd typically use a query on a GSI to avoid a scan.

```
import Helpers from './helpers.js'
import Schema from '../schemas/Schema.0.0.15.js'

export default {
    version: '0.0.15',
    description: 'Update Member GS2',
    schema: Schema,

    async up(db, migrate) {
        let start = null
        do {
            let items = await db.scan('Member', {limit: 1000, start})
            for (let item of items) {
                await db.update('Member', item, {updateIndexes: true})
            }
            start = items.start
        } while (start)

        start = null
        do {
            let items = await db.scan('User', {limit: 1000, start})
            for (let item of items) {
                await db.update('User', item, {updateIndexes: true})
            }
            start = items.start
        } while (start)
    },

    async down(db, migrate) {
        // No revert
    }
}

```

What kinds of examples would you like to see?
